### ISSUE TITLE
Vector synthesis plugin?

### ISSUE THREAD

**Issue by mxmilkiib** [ryukau/LV2Plugins#29]
https://en.m.wikipedia.org/wiki/Vector_synthesis
https://youtu.be/Zj61tEKx4ro

Kinda related;
https://youtu.be/ufBeolCjWSw - "poly" synthesis
https://youtu.be/rtR63-ecUNo - visual vector synthesis

**Comment by ryukau** [comment]
I think there's 2 meaning of vector synthesis here.

The one is described on Wikipedia, which is just crossfade. I think this is not interesting without combining some modulation.

Another is drawing Lissajous curve. If you are talking about second one, there's 2 direction: Image to sound, and sound to image. I'm not interested in sound to image conversion because finding nice sound with fancy Lissajous curve is artistic process.

Poly synthesis video is probably using the methods in following paper.

- [EFFICIENT ANTI-ALIASING OF A COMPLEX POLYGONAL OSCILLATOR](http://dafx.de/paper-archive/2017/papers/DAFx17_paper_100.pdf)

**Comment by magnetophon** [comment]
I'm working on a monophonic synth with the following features:
- 4 oscillators in double crossfade configuration, so VS in the wikipedia sense.
- each osc is a [CZ style](https://forum.pdpatchrepo.info/topic/5992/casio-cz-oscillators) osc, where you can choose between 9 waveforms
- each osc is phase-locked to a single master
- but the phase can be changed and modulated
- each osc can crossfade between octaves
- the phase signal feeding the CZ oscs can be passed trough a set of lowpass-filters
  this gives a similar yet very different sound character. for example: when you filter the phase input of a sine-wave, it creates harmonics and you can kind of hear the character of the filter. 
  - there are a couple of lowpass-types (oberheim/ms20/moog(WIP)) that can be dynamically mixed, but the sum of the mix is always 1
- the synth is fully stereo: each parameter has a main slider and a left-right difference slider
- each parameter (and it's l-r-diff brother) can be modulated by 4 envelopes and 4 lfo's
- the number of envelopes and lfos can be easily changed at compile time


Quick [audio demo](https://magnetophon.nl/sounds/magnetophon/digiDrie.mp3).
Here's [the source](https://github.com/magnetophon/faustExperiments/blob/globalFilter/lastNote.dsp).

@ryukau This synth really needs a proper UI, ideally one where the modulation is visualized like in Helm and Surge.

**Comment by ryukau** [comment]
@magnetophon If you are OK with the generic style used in this repository, I'll work on it. However, keep in mind that visualization will have some limitations depends on the number of values that have to send from DSP to UI. Displaying phase of LFO/envelope is feasible, but component like oscilloscope probably degrades performance, because it requires access to audio buffer.

**Comment by magnetophon** [comment]
@ryukau Cool!  Of course I'm OK with your style, it looks great!

To me the main point is making the plugin simpler to use.
The sliders of the modulation mixers take up too much screen space and attention. 
Helm and surge solve this nicely.

A scope is unneeded, performance is way more important.

Do you have any requests for features or maybe for dropping features?
I was going to work on:
- making the envelopes more flexible, for example a choice to either re-trigger on each note, or only at the start of a legato phrase.
- adding retrigger options to the LFO's
- adding tempo sync to the LFO's.
- adding tempo sync to the envelopes.
- making the octave slider more versatile: add an option to always "transpose" the note to the same frequency range.
  One usage example is: adding a sine wave that is always in the bass range, no matter what note you play. 


Some of these functions (LFO and env related) are already available in your plugins, so maybe I should skip working on those and leave them to you.

Also, the monophonic handeling is done in a rather clunky way in faust at the moment. 
Probably best to use your code for that.

Have you used faust before?

**Comment by ryukau** [comment]
It's almost midnight (0:00 in 24h) here, so I'll try the synth tomorrow.

I played with Faust like a year or 2 ago. Only learned basic syntax, but I can make environment.  Just in case, I'd like to know the Faust version you are using. Probably latest master branch?

If I understand correctly, proper tempo sync isn't possible for now. See [this issue](https://github.com/DISTRHO/DPF/issues/235) in DPF. Basically it goes out of sync when tempo is changed. I can work to patch LV2 and VST 2 code path, but I'm not sure about other formats.

**Comment by magnetophon** [comment]
There's just been a new [faust release](https://github.com/grame-cncm/faust/releases/tag/2.27.1), so I'm using that.

I'm compiling with `faust2jack -t 0 -time -midi lastNote.dsp`.
Also, make sure you get the [correct branch](https://github.com/magnetophon/faustExperiments/blob/globalFilter/lastNote.dsp)

For tempo sync, I was going to use [this code](https://github.com/sonejostudios/SuperBeatRepeater/blob/1.1/SuperBeatRepeater.dsp#L114). Thanks for publishing it, @sonejostudios!
I haven't tried it out yet, so I don't know how it behaves with tempo-changes.

I think LV2 and VST 2 will cover most use cases.

**Comment by ryukau** [comment]
The synth sounds pretty good. I like it.

Feature requests:
- Phase smoothing time parameter.
- Modulation between oscillators, with gain affected by 2D crossfade.
- Oscillator pan with LFO modulation.
- Key follow modulation.

Issues:
- Oscillator cause glitch when `master->oct` and `oct` are both -4, and modulated to go lower.
- Massive aliasing when `index` is set to max, and modulated to go higher.
- Filters (Allpass, ms20, oberheim) aren't working. When turning up level, they output glitchy sound.
- Note-on while release cause pop noise.

I have an impression that modulated values aren't bounded. I personally like those kind of aliasing and glitches, however it might be better to add some conditions.

```faust
// hardclip may be better name in context of Faust.
clamp(x, low, high) = (x, low : max), high : min;

index = clamp(index + indexModulation, minIndex, maxIndex);
oct = clamp(oct + octModulation, minOct, maxOct);
```

I have a mitigation to note-on while release problem. So if it's hard to solve on Faust, please let me know.

> - making the octave slider more versatile: add an option to always "transpose" the note to the same (user defined) frequency range.
One usage example is: adding a sine wave that is always in the bass range, no matter what note you play.

If I understand correctly, I used the word "fixed" to describe this behavior in SyncSawSynth.

I think retrigger is better implemented in Faust. However, I'm not sure about how Faust generated code handles MIDI notes and time information. I'll test tempo sync and monophonic note handling later.

**Comment by magnetophon** [comment]
> The synth sounds pretty good. I like it.

Thanks!

> Feature requests:
> 
> * Phase smoothing time parameter.

Could you explain a bit more?

> * Modulation between oscillators, with gain affected by 2D crossfade.

You mean hooking up the 2 crossfade sliders to the envelopes and LFO's?  That is on my TODO list.

> * Oscillator pan with LFO modulation.

Currently there is one set of oscillators hard-panned left and one set hard-panned right.
Stereo is created by changing the settings of the oscillators and/or filters.
Isn't that enough options to create an interesting stereo image?

> * Key follow modulation.

You mean "key follow" for the filters?
Anything else?

> 
> Issues:
> 
> * Oscillator cause glitch when `master->oct` and `oct` are both -4, and modulated to go lower.

Yeah, I should bound the modulation.
The number of octaves is configurable, should I change it?

> * Massive aliasing when `index` is set to max, and modulated to go higher.

Yup, need to bound the modulation.

> * Filters (Allpass, ms20, oberheim) aren't working. When turning up level, they output glitchy sound.

They filter the phase input of the oscillators, and as such only sound good when there is no "dry signal" left.
The level controls feed a "fallback mixer":
- When the sum of the (abs) filter levels is between 0 and 1, you get a mix of the dry signal and the filter.
- When the sum is 1 or higher, you get the mix of filters, scaled down to a total volume of 1.

The allpass always sounds glitchy on it's own, but can be cool when mixed in with the other filters.
I think I'll replace it with a moog. 

So, to get a cool filter sound,like in the mp3 above, do the following:
- Set the ms20 level to 1
- set the L-R offset of the filter freq and Q to minimum  (this should be the default, but I haven't implemented that yet)
- play with the main freq and Q sliders and modulation
- optional: mix in about negative 0.5 of oberheim, (currently, you can only get negative mix amounts via the envelopes )

In general, always make sure the sum of all filter levels is 1 or higher, otherwise unfiltered signal will be mixed in with filtered signal, which sounds glitchy.

> * Note-on while release cause pop noise.

I'll investigate.
It's probably because I reset the phase for each note and I need to make sure that doesn't happen when there is still a note playing.
I also need a switch to make the master osc free-running.

> 
> I have an impression that modulated values aren't bounded. I personally like those kind of aliasing and glitches, however it might be better to add some conditions.

Yeah, most modulations aren't bounded yet. WIP.


> I have a mitigation to note-on while release problem. So if it's hard to solve on Faust, please let me know.

Does it solve the phase reset problem I described above, or did you have a different problem?

> If I understand correctly, I used the word "fixed" to describe this behavior in SyncSawSynth.

No, this is different:
The frequencies you hear are determined by both the note you play and the "target frequency slider".
You will always get the correct note, but it will be shifted into the octave around the target frequency.

Say the target freq is 220 Hz.
You play a 440Hz midi note, you hear 220 Hz
play 110, also hear 220
play 150, hear a mix of 150 and 300
play 600, also hear a mix of 150 and 300


Probably easiest to try it for yourself:
```
wget https://github.com/magnetophon/faustExperiments/blob/globalFilter/SubSynth.dsp
faust2jack -time -midi -nvoices 1 SubSynth.dsp && ./SubSynth &

```

> I think retrigger is better implemented in Faust.

OK,I'll work on it.

>However, I'm not sure about how Faust generated code handles MIDI notes and time information. I'll test tempo sync and monophonic note handling later.

Faust currently has [quite bad monophonic note handling](https://github.com/grame-cncm/faust/issues/252), that's why I worked around it in faust code.



Thanks for looking in to this! I'm quite excited about our little project!

**Comment by mxmilkiib** [comment]
[sisco.lv2](https://github.com/x42/sisco.lv2) is a great scope, and there is one [forthcoming](https://github.com/sadko4u/lsp-plugins/commits/oscilloscope) from LSP. A sidenote; the [MOOC](https://www.kadenze.com/courses/real-time-audio-signal-processing-in-faust/info) videos with @orlarey and @rmichon are great as an intro/refresher. Maybe [some day](https://github.com/grame-cncm/faust/issues/435) metadata/CV ports will come to FAUST.

**Comment by ryukau** [comment]
> - Phase smoothing time parameter.

I'm pretty sure `si.smooth` is used somewhere to smooth the phase. I'm thinking about code like below:

```faust
phaseSmoothSeconds = hslider(...); // User input.
uiPhase = uiPhase: si.smooth(tau2pole(phaseSmoothSeconds));
```

> - Modulation between oscillators, with gain affected by 2D crossfade.

The synth has modulation matrix. The idea is to add oscillator output from 1 frame earlier to the modulation matrix.

I'm thinking about code like below. I couldn't come up with Faust code, so the code is C++. In Faust, this is probably the use case of recursion `~`.

```c++
float osc1Out;

float process(/* ... */) {
  modulationToOsc1Phase
    = env1Out * env1ModToPhase1
    + env2Out * env2ModToPhase1
    + /* ... */
    + osc1Out * osc1ModToPhase1
    + osc2Out * osc2ModToPhase1
    + /* ... */;

  // ...

  osc1Out = CZosc.process(phase + modulationToOsc1Phase, ...) * crossfaderGain;

  // ...

  oscOut = osc1Out + osc2Out + osc3Out + osc4Out;

  // ...
}
```

`*Out` are output of modulators. `*ModToPhase1` are user inputs of modulation amount. In this case, it's phase modulation. But oscillator output may be used to modulate `index`, `oct`, filter cuttoff and resonance.

> - Oscillator pan with LFO modulation.

I haven't noticed there's L-R offset. Sorry for that. This request is then unnecessary.

> - Key follow modulation.

Key follow for index is also nice to have. Use case is to reduce the modulation amount at higher notes.

> The number of octaves is configurable, should I change it?

If it doesn't cause glitch, more lower octaves are nice to have.

> Does it solve the phase reset problem I described above, or did you have a different problem?

It solves the phase reset problem. However, slide from release is not possible with it.

My synths are using following method to reduce pop noise in case of note-on while playing all voices.

1. Receives note-on while playing all voices.
2. Pick up most quiet note.
3. Render fade out of the picked up note to a ring buffer.
4. Set new parameters to the picked up note.
5. Return to processing loop and start adding the content of ring buffer.

This method is a kind of brute forcing, so it can be heavy. The image below shows the note envelope on above method.

![max_poly_note_on](https://user-images.githubusercontent.com/6243462/88301751-fc6f2e80-cd3f-11ea-9f3a-3a1766ba3290.png)

The idea of SubSynth is new to me. Thanks for the explanation! I guess this has really nice applications.

I also think global pitch control (transposing octave, semitone, cents) is nice to have. I could provide the one used in CubicPadSynth.

---

@mxmilkiib Session 5 of the course seems interesting. Thanaks for the information.

**Comment by magnetophon** [comment]
Some general updates:

- the 2D crossfade is now modulateable.
- Oscillators now alias a lot less
- bug fix for phase: now works correct for all octaves.

> > * Phase smoothing time parameter.

I understand now.

I want to make macro sliders available as mod-sources, and those macro sliders will have a smoothing time parameter.
Possibly a separate one for up and down. 

Would that be an OK alternative to your request?
 
> > * Modulation between oscillators, with gain affected by 2D crossfade.

I added PM. Thanks for the idea, sounds wicked!
I used the signal before it goes into the crossfade, hope that's OK!
Maybe I'll add PM from the mixed oscillators as an option as well. 

>  In this case, it's phase modulation. But oscillator output may be used to modulate `index`, `oct`, filter cuttoff and resonance.
> 

That will explode the number of parameters, are you sure you want that?
Will you e able to make a GUI that makes the synth understandable?

> > * Key follow modulation.
> 
> Key follow for index is also nice to have. Use case is to reduce the modulation amount at higher notes.

I tested the maximum amount of index each frequency could have without too much aliassing.
The incoming phase signal is used to calculate the frequency at that moment, and the incoming index parameter is scaled so it doesn't alias.

I will still add key follow as a mod source, but possibly via the macro system, so the amount of parameters doesn't grow too much.

There are more than 700 sliders already!  :grin:  

> If it doesn't cause glitch, more lower octaves are nice to have.

OK, we can go down 8 and up 4 octaves now.
Enough?
 
> > Does it solve the phase reset problem I described above, or did you have a different problem?

I made the synth more intelligent when it comes to resetting the phase:
- you can switch on free-running, so the phase never gets reset.
- Without free-running, the phase is only reset when the amp envelope is below 0.1.

> It solves the phase reset problem. However, slide from release is not possible with it.

What do you mean by slide from release?
 
> This method is a kind of brute forcing, so it can be heavy. The image below shows the note envelope on above method.

The synth is already quite heavy: it uses 42% CPU on my old thinkpad T520.
What hardware are you running it on, and how heavy is it for you?
  
> The idea of SubSynth is new to me. Thanks for the explanation! I guess this has really nice applications.

Still todo.
> 
> I also think global pitch control (transposing octave, semitone, cents) is nice to have. I could provide the one used in CubicPadSynth.

Also still todo.

**Comment by ryukau** [comment]
This is getting long, so please let me know if I'm missing topics. Also if you have idea for better format, I'm open to it.

## Reply to last comment
I tested the latest commit and it sounds like most of the rough edges are gone now. Especially, filters feels clean now.

> I want to make macro sliders available as mod-sources, and those macro sliders will have a smoothing time parameter.
> Possibly a separate one for up and down.
>
> Would that be an OK alternative to your request?

Would you elaborate on what macro means here?

The point is to have a control to change the character of sound when the value is modulated by DAW automation, or by hand through GUI. I'm OK with if the smoothing time is global, and affects not to only phase but to all parameters using smoothing.

> I added PM. Thanks for the idea, sounds wicked!
> I used the signal before it goes into the crossfade, hope that's OK!
> Maybe I'll add PM from the mixed oscillators as an option as well.

Thanks for this addition, it sounds great! Combined with filters and lower octaves you just added, it's really nice.

Some output with PM: https://drive.google.com/file/d/1Ppus-q512l8vSRVcMVSuptmIkvU0k-3_/view?usp=sharing

> That will explode the number of parameters, are you sure you want that?
> Will you e able to make a GUI that makes the synth understandable?

I'm planning to make UI like the image below.

![lastNote_matrix_plan](https://user-images.githubusercontent.com/6243462/88640585-11114500-d0f9-11ea-84fe-4c185e2f55cb.png)

I got the idea from [Sytrus](https://www.image-line.com/support/flstudio_online_manual/html/plugins/Sytrus.htm). The image in [The Modulation Matrix section](https://www.image-line.com/support/flstudio_online_manual/html/plugins/Sytrus.htm#Sytrus_modmatrix) is the default size.

Probably I'll put the matrix on a dedicated tab. It's better to get a global picture of modulation at a glance.

> I tested the maximum amount of index each frequency could have without too much aliassing.
> The incoming phase signal is used to calculate the frequency at that moment, and the incoming index parameter is scaled so it doesn't alias.

Is [this](https://github.com/magnetophon/faustExperiments/blob/ccea33b99a404e553299ee416ab83b6eb9c08751/lastNote.dsp#L1072) part? I think it can be simplified a bit by eliminating `ma.SR`.

```faust
diffHold(fund) = // Former index2freq.
  ba.sAndH(df, abs(df) < 0.5) with { df = fund - fund'; };

indexAA(index, fund) = // Anti Alias => lower the index for higher freqs
  index * (1 - tuned) with {
    tuned = (diffHold(fund) * 8 - 1 / 32) : max(0) : min(1);
  };

resAA(res, fund) = res : max(0) : min(1 / (4 * diffHold(fund)));
```

If the input of `index2freq` is normalized phase in [0, 1), wrap around handling is missing I guess.

```faust
df = if(fund < fund', 1, 0) + fund - fund';
```

Current implementation has its own character, so maybe it's better to leaving it as is.

> There are more than 700 sliders already! grin

I guess we can go further. 😎

> OK, we can go down 8 and up 4 octaves now.
> Enough?

Yes, this is good. Thanks again for addition!

> What do you mean by slide from release?

I meant pitch sliding. In lastNote parameter, portamento is the equivalent. I'd like to have a option to toggle portamento from release.

TrapezoidSynth has implementation of this behavior. To try it, change Slide option from `Sustain` to `Always`.

> The synth is already quite heavy: it uses 42% CPU on my old thinkpad T520.
> What hardware are you running it on, and how heavy is it for you?

I'm using Ryzen 7 1700. qjackctl indicates DSP load around 36% here.

The method I mentioned will doubles or even triples the DSP load in a cycle, depending on the buffer size. So it's better to think my method as a last resort.

If you are on to free up some CPU, I can think of following possibility.

- Moving parameter mapping to UI. `faust` command provides `--one-sample` option. So I can provide some mapping and smoothing for heavy one.
- Using SIMD. I'm not sure about the compile options Faust using, but it might become faster by explicitly specifying `-mavx` or `-mavx2`.
- Omittng L-R offset. While it's cumbersome, user can launch 2 instances of same synth and make them stereo.

To elaborate the first point, it will look like following.

```c++
param = getUIValue();

for (/* loop inside audio buffer */) {
  faustDsp.param = smoother.process(param);
  output = faustDsp.process();
}
```

## Issue
Another issue: Notes are ignored with certain procedure.

Steps to reproduce:

1. Trigger note A.
2. While holding down note A, trigger another note B. Note B must be higher than note A.
3. Release note A.
4. Trigger note A again. In my environment, this note is ignored.

On step 4, notes that lower than note B are also ignored. The synth starts accepting those notes after certain times are spent.

## Tempo Sync Related Informations
On tempo sync, I made a [patch to DPF](https://github.com/DISTRHO/DPF/pull/218). With this patch, I can send the following informations:

- Fraction part of beat at the start of current DSP cycle.
- Current beat in a bar.
- Time signature. Upper numeral may be sufficient for tempo sync.
- Or any other information provided by `BarBeatTick` in DPF ([link toreference](https://distrho.github.io/DPF/structTimePosition_1_1BarBeatTick.html)).

Following is an example C++ code to calculate remaining frame count to the start of next beat:

```c++
// beatFraction is in [0.0, 1.0).

double framesToNextBeat = beatFraction == 0.0
    ? 0
    : framesPerBeat * (1.0 - beatFraction);
```

`framesToNextBeat` can be either integer or float.

These informations are only available on LV2, VST 2. So if you'd like to target JACK standalone, it requires different code path which uses MIDI clock for tempo sync.

To add this, it seems possible to provide same information on VST 3 ([link to reference](https://steinbergmedia.github.io/vst3_doc/vstinterfaces/structSteinberg_1_1Vst_1_1ProcessContext.html)). I'll test this later.

## C++ Code Generation
Currently, I'm planning to use `faust` command to generate C++ code. In order to integrate, I have to solve following problems:

- Generated code refers to several C++ headers in `faust/architecture/faust`.
- There seems no way to strip out the GUI method call in generated code.
- There seems no way to change parameter name in generated code.

To elaborate second and last point, `faust` command generates C++ code like below.

```c++
class mydsp : public dsp {
 private:

	FAUSTFLOAT fHslider0;

 public:

	virtual void buildUserInterface(UI* ui_interface) {
		ui_interface->addHorizontalSlider("gain", &fHslider0, 0.5f, 0.0f, 1.0f, 0.00999999978f);
	}
};
```

What I'd like to get is code like below.

```c++
class mydsp : public dsp {
 public:
	FAUSTFLOAT gain;
};
```

Basically I'd like to replace generic names like `fHslider0` to `gain` or some proper name provided by first argument of Faust GUI functions.

I'm planning to write some python to transform the C++ code. If you know a better way, please let me know.

**Comment by magnetophon** [comment]
> Also if you have idea for better format, I'm open to it.

I made [a repo](https://github.com/magnetophon/DigiDrie) where you have access.
Let me know if you have ideas for a better name!

> Would you elaborate on what macro means here?

I implemented macros in the new repo.
Think of them as presets that you can crossfade between.
At the moment the modulation sources for crossfading are:
- a fader on the macros page
- the envelopes
- the lfo's, shifted so they are in the range 0..1.

I want to add as macro mod soures:
- note
- veloity
- mod-wheel
- aftertouch

To make the macro system usable, we need a function in the gui to copy paste settings for example:
- copy from main settings to macro 0
- copy from macro 4 to macro 1.

> The point is to have a control to change the character of sound when the value is modulated by DAW automation, or by hand through GUI. I'm OK with if the smoothing time is global, and affects not to only phase but to all parameters using smoothing.

I think you can get the effect you want by doing the following:
- Start with the default settings
- go to tab A -> main -> CZ and in there set parameter  phase -> modulation sources -> macros -> 0 -> phase to the value you want
- go to tab macros and play with macro 0, uptime 0 and downtime 0

> I'm planning to make UI like the image below.

I think it's cool to have a modulation matrix like that, it gves you a great detailed overview.
But to make the synth really usable, I think we need to visualise the modulations, like in helm and surge.
That also means we need a main page where you see just one knob/sloder per parameter, and one knob/slider per mod-source.
Have you played with helm and surge?


> Is [this](https://github.com/magnetophon/faustExperiments/blob/ccea33b99a404e553299ee416ab83b6eb9c08751/lastNote.dsp#L1072) part? I think it can be simplified a bit by eliminating `ma.SR`.

Well the whole point of the thing is to have lower index at higher frequencies.
I need the SR to know the freqency of the phase signal.

Am I missing something?

> Current implementation has its own character, so maybe it's better to leaving it as is.



> > What do you mean by slide from release?
> 
> I meant pitch sliding. In lastNote parameter, portamento is the equivalent. I'd like to have a option to toggle portamento from release.
> 
> TrapezoidSynth has implementation of this behavior. To try it, change Slide option from `Sustain` to `Always`.

Can you make an issue an the repo, so I don't forget?

> If you are on to free up some CPU, I can think of following possibility.
> 
> * Moving parameter mapping to UI. `faust` command provides `--one-sample` option. So I can provide some mapping and smoothing for heavy one.

Please explain.

> * Using SIMD. I'm not sure about the compile options Faust using, but it might become faster by explicitly specifying `-mavx` or `-mavx2`.

I didn't find those options on my version of faust.
I did try with -fm def, ut that didn't ompile.
I opened an issue [here](https://github.com/grame-cncm/faust/issues/472)

> * Omittng L-R offset. While it's cumbersome, user can launch 2 instances of same synth and make them stereo.

I can make a light version of the plugin as well, but for the main version, this is one of the coolest features, even though it doules the CPU usage.
You can make really interesting stereo effects since all the oscillators are locked to one master.

> ## Issue
> Another issue: Notes are ignored with certain procedure.

I can reproduce.
The way notes are handeled currently is a cludge in faust.
I think we need tou use your code for this.

>
> ## Tempo Sync Related Informations
> On tempo sync, I made a [patch to DPF](https://github.com/DISTRHO/DPF/pull/218). With this patch, I can send the following informations:

I don't follow. I can not really read C++ code, sorry.
Will DPF send tempo and position info to my faust code?

> ## C++ Code Generation
> Currently, I'm planning to use `faust` command to generate C++ code. In order to integrate, I have to solve following problems:
>
> I'm planning to write some python to transform the C++ code. If you know a better way, please let me know.

This should be possible, I think, but I never used any of the export to source options of faust. I just use faust to make binarys directly.
Maye ask on [the faust slack](https://app.slack.com/client/T0Z8G8011/C0ZA8T51Q?cdn_fallback=0)?

**Comment by sletz** [comment]
For C++ code generation, this [faustpp](https://github.com/jpcima/faustpp) project may help.

**Comment by sletz** [comment]
And for C++ code optimisation see:
- https://faustdoc.grame.fr/manual/optimizing/
- https://github.com/grame-cncm/faust/tree/master-dev/tools/benchmark

**Comment by ryukau** [comment]
To @magnetophon:

> I made a repo where you have access.

For some reply, I'll open issues on there.

> Let me know if you have ideas for a better name!

As you can see in this repo, I'm not a person good at naming things. I'll leave it to you ;)

> I think you can get the effect you want by doing the following:
>
> - Start with the default settings
> - go to tab A -> main -> CZ and in there set parameter phase -> modulation sources -> macros -> 0 -> phase to the value you want
> - go to tab macros and play with macro 0, uptime 0 and downtime 0

It did the trick. Thank you!

> I think it's cool to have a modulation matrix like that, it gves you a great detailed overview.
> But to make the synth really usable, I think we need to visualise the modulations, like in helm and surge.
> That also means we need a main page where you see just one knob/sloder per parameter, and one knob/slider per mod-source.
> Have you played with helm and surge?

I'll try helm and surge.

> > - Using SIMD. I'm not sure about the compile options Faust using, but it might become faster by explicitly specifying -mavx or -mavx2.
>
> I didn't find those options on my version of faust.
> I did try with -fm def, ut that didn't ompile.
> I opened an issue here

`-mavx` and `-mavx2` are options for C++ compilers (g++ and clang++). I'll try Faust benchmark tools (thanks to @sletz) for this one.

> I can reproduce.
> The way notes are handeled currently is a cludge in faust.
> I think we need tou use your code for this.

I'll provide note handling on C++.

---

@sletz: Thanks for the information! I'll look into them.
